##**Function to convert a Microsoft Excel Epoch date to a "YYYY-MM-DD" date Format.**

##**Books borrowed and returned were in 2023.**

In [ ]:
import pandas as pd
import datetime

df1=pd.read_csv("/content/books (1).csv")

# df1 = pd.read_csv("books(1).csv")
df = pd.read_csv("bookloans (1).csv")

df1.isna()

def excel_to_date(excel_date: int) -> str:
    """
    Convert a date represented as the number of days since the Excel epoch to a string in the format 'YYYY-MM-DD'.

    Parameters:
        excel_date (int): The number of days since the Excel epoch (December 30, 1899) to be converted.

    Returns:
        str: A string in the format 'YYYY-MM-DD' representing the date corresponding to the input number of days.
    """
    # Define the starting date of the Excel epoch
    excel_epoch_start = datetime.datetime(1899, 12, 30)

    # Calculate the Python date by adding the number of days to the Excel epoch start
    py_date = excel_epoch_start + datetime.timedelta(days=excel_date)

    # Format the Python date as a string in 'YYYY-MM-DD' format
    return py_date.strftime('%Y-%m-%d')

# Example usage
# print(excel_to_date(45299))
# print(excel_to_date(45553))


# Define your header names as a list
header_names = ["book_number", "member_number", "date_of_loan" ,"date_of_return"]

# Read the CSV file with the specified header names
df = pd.read_csv("/content/bookloans (1).csv", header=None, names=header_names)

# print(df)


# converting date
date = ""
for index,value in enumerate(df["date_of_loan"]):
    date = excel_to_date(value)
    df["date_of_loan"][index] = date

# Convert Excel dates_of_return to 'YYYY-MM-DD' format
df["date_of_return"] = df["date_of_return"].apply(excel_to_date)

# print (df)


# Filter the DataFrame to keep only rows from the year 2023
df_2023 = df[(df["date_of_loan"] >= "2023-01-01") & (df["date_of_loan"] <= "2023-12-31")]
df_2023 = df[(df["date_of_return"] >= "2023-01-01") & (df["date_of_return"] <= "2023-12-31")]

# Display the filtered DataFrame
print(df_2023)


      book_number  member_number date_of_loan date_of_return
0              16            126   2023-01-01     2023-01-18
1              31            192   2023-01-01     2023-01-19
2              57            199   2023-01-01     2023-01-21
3             100            140   2023-01-01     2023-01-15
4             100             39   2023-01-01     2023-01-08
...           ...            ...          ...            ...
2204           49            117   2023-12-30     2023-12-31
2205           62            130   2023-12-30     2023-12-31
2206           63             19   2023-12-30     2023-12-31
2207           94            178   2023-12-30     2023-12-31
2208           99             53   2023-12-30     2023-12-31

[2158 rows x 4 columns]


# **Task 1**

## **PART I**

### **books bought multiple times**

In [ ]:
# Group the DataFrame by "book_number" and count the occurrences of each book
book_counts = df.groupby("book_number").size().reset_index(name="purchase_count")

# Filter the DataFrame to only include books that have been purchased more than once
multi_purchase_books = book_counts[book_counts["purchase_count"] > 1]

# Display the books that have been purchased multiple times
print("Books purchased multiple times:")
print(multi_purchase_books)





Books purchased multiple times:
     book_number  purchase_count
0              1              22
1              2              21
2              3               8
3              4               4
4              5              22
..           ...             ...
135          136              13
136          137               9
137          138              21
138          139               8
139          140               6

[138 rows x 2 columns]


**Top 5 most popular books**

In [ ]:
# Sort the DataFrame by the purchase count in descending order
book_counts_sorted = book_counts.sort_values(by="purchase_count", ascending=False)

# Display the top 5 most sold books
print("Top 5 most sold books:")
print(book_counts_sorted.head(5))

Top 5 most sold books:
     book_number  purchase_count
84            85              28
118          119              28
37            38              28
71            72              27
60            61              27


### **The 5 most popular books are:**
```
Book Number       Book Title                        Total bought
1. 85       |     The Great Indian Novel           |     28
2. 118      |     One                              |     28
3. 37       |     The Veteran                      |     28
4. 71       |     All The President'S Men          |     27
5. 60       |     Data Structures Using C & C++    |     27
```



**least popular books**

In [ ]:
# Sort the DataFrame by the purchase count in descending order
book_counts_sorted = book_counts.sort_values(by="purchase_count", ascending=True)

# Display the top 5 most sold books
print("5 least popular books are:")
print(book_counts_sorted.head(5))

5 least popular books are:
     book_number  purchase_count
122          123               1
12            13               1
34            35               2
56            57               2
58            59               2


### **The 5 least popular books are:**

```
Book Number      Book Title                         Total bought
1. 123     |     Selected Short Stories            |     1
2. 13      |     Birth Of A Theorem                |     1
3. 35      |     Surely You'Re Joking Mr Feynman   |     2
4. 57      |     Textbook Of Economic Theory       |     2
5. 59      |     Learning Opencv                   |     2
```



## **PART II**

# **book appears on the loan list in 2023.**   
**reverse order of their popularity.**

**all rows dislayed**

In [ ]:
import pandas as pd
import datetime

# Function to convert Excel date to 'YYYY-MM-DD' format
def excel_to_date(excel_date: int) -> str:
    excel_epoch_start = datetime.datetime(1899, 12, 30)
    py_date = excel_epoch_start + datetime.timedelta(days=excel_date)
    return py_date.strftime('%Y-%m-%d')

# Define header names for loans DataFrame
header_names = ["book_number", "member_number", "date_of_loan" ,"date_of_return"]

# Read the loans CSV file with the specified header names
df_loans = pd.read_csv("/content/bookloans (1).csv", header=None, names=header_names)

# Convert Excel dates to 'YYYY-MM-DD' format
df_loans["date_of_loan"] = df_loans["date_of_loan"].apply(excel_to_date)
df_loans["date_of_return"] = df_loans["date_of_return"].apply(excel_to_date)

# Filter loans DataFrame to keep only rows from the year 2023
df_loans_2023 = df_loans[(df_loans["date_of_loan"] >= "2023-01-01") & (df_loans["date_of_loan"] <= "2023-12-31")]

# Count the number of times each book is borrowed in 2023
books_borrowed_2023 = df_loans_2023["book_number"].value_counts().reset_index()
books_borrowed_2023.columns = ["book_number", "maximum duration of each Book rented"]

# Set display options to show all rows
pd.set_option('display.max_rows', None)

print("Number of days for which each book was borrowed in 2023:")
# Display the count of borrowings for each book in 2023
print(books_borrowed_2023)


Number of days for which each book was borrowed in 2023:
     book_number  maximum duration of each Book rented
0            119                                    28
1             38                                    28
2             85                                    28
3             61                                    27
4             72                                    27
5             43                                    25
6             51                                    25
7            132                                    25
8             48                                    25
9            105                                    25
10            66                                    25
11            94                                    25
12           112                                    24
13            78                                    24
14           134                                    24
15           128                                    24
16      

In 2023, the library counted how many times each book was borrowed to see which ones were popular among readers. This helps the library decide which books to keep based on how often they are checked out, showing which ones people enjoy and find valuable.

# **Task 2**
###**Types of Genres and movies related to each Genre**

In [ ]:
import pandas as pd
import datetime

# Define header names for the book loans data
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Load the datasets
updated_book_loans_path = '/content/bookloans (1).csv'
updated_books_path = '/content/books (1).csv'

# Read the book loans data with headers
updated_book_loans = pd.read_csv(updated_book_loans_path, header=None, names=header_names)

# Read the books data
updated_books = pd.read_csv(updated_books_path)

# Define the function to convert Excel serial date to year
def serial_date_to_string(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=s)
    return (excel_epoch + delta).year

# Apply conversion to date_of_loan column to filter for 2023
updated_book_loans['year_of_loan'] = updated_book_loans['date_of_loan'].apply(serial_date_to_string)

# Filter for the year 2023
loans_2023 = updated_book_loans[updated_book_loans['year_of_loan'] == 2023]

# Join the filtered loan data with the books data
loans_with_genre = loans_2023.merge(updated_books, left_on='book_number', right_on='Number')

# Count occurrences of each genre
genre_counts = loans_with_genre['Genre'].value_counts().reset_index()
genre_counts.columns = ['Genre', 'Number of Books']

# Sort by count in descending order
sorted_genre_counts = genre_counts.sort_values(by='Number of Books', ascending=False)

print("popularity report of all genres of books in 2023")
sorted_genre_counts


popularity report of all genres of books in 2023


Genre  Number of Books
0      Fiction              717
1  Non-fiction              591
2         Tech              396
3      Science              153
4   Philosophy              101

The  cell displays the categories in the library long with the total numbers of books within each category.

# **Task 3**
### **Statistics on loans and late loans are lost.**

 *• Number of loans.*

*• Number of days borrowed.*

*• Number of loans returned late.*

*• Number days late.*

In [ ]:
import pandas as pd
import datetime

# Define your header names as a list
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Read the CSV file with the specified header names
updated_book_loans = pd.read_csv('/content/bookloans (1).csv', header=None, names=header_names)

# Define the function to convert Excel serial date to datetime
def serial_date_to_datetime(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=s)
    return excel_epoch + delta

# Apply conversion to the loan and return dates
updated_book_loans['date_of_loan'] = updated_book_loans['date_of_loan'].apply(serial_date_to_datetime)
updated_book_loans['date_of_return'] = updated_book_loans['date_of_return'].apply(serial_date_to_datetime)

# Calculate total number of loans
total_loans = len(updated_book_loans)

# Calculate total number of days borrowed, ensure no negative days are counted
updated_book_loans['days_borrowed'] = (updated_book_loans['date_of_return'] - updated_book_loans['date_of_loan']).dt.days
updated_book_loans['days_borrowed'] = updated_book_loans['days_borrowed'].apply(lambda x: max(x, 0))
total_days_borrowed = updated_book_loans['days_borrowed'].sum()

# Calculate number of loans returned late
updated_book_loans['is_late'] = updated_book_loans['days_borrowed'] > 14
total_books_returned_late = updated_book_loans['is_late'].sum()

# Calculate total days late
updated_book_loans['days_late'] = updated_book_loans['days_borrowed'] - 14
updated_book_loans['days_late'] = updated_book_loans['days_late'].apply(lambda x: x if x > 0 else 0)
total_days_late = updated_book_loans['days_late'].sum()

# Output the results
print("\033[1mTotal number of loans:\033[0m" +"          |  "+ str(total_loans))
print("\033[1mTotal number of days borrowed:\033[0m"+"  |  "+ str(total_days_borrowed))
print("\033[1mTotal loans returned late:\033[0m" +"      |  "+ str(total_books_returned_late))
print("\033[1mTotal days late:\033[0m" +"                |  "+ str(total_days_late))



Total number of loans:          |  2209
Total number of days borrowed:  |  22135
Total loans returned late:      |  613
Total days late:                |  2144


The calculations show that in total, there were 2,209 loans made from the library, Gather to 22,135 days borrowed. Among these, 613 loans were returned late, resulting in a total of 2,144 days overdue.

## **a) The average number of days that a book was borrowed.**

### ***calculating the average number of days a book was borrowed across all loans, rather than for each individual book.***

In [ ]:
import pandas as pd
import datetime

# Define your header names as a list
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Read the CSV file with the specified header names
updated_book_loans = pd.read_csv('/content/bookloans (1).csv', header=None, names=header_names)

# Define the function to convert Excel serial date to datetime
def serial_date_to_datetime(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=s)
    return excel_epoch + delta

# Apply conversion to the loan and return dates
updated_book_loans['date_of_loan'] = updated_book_loans['date_of_loan'].apply(serial_date_to_datetime)
updated_book_loans['date_of_return'] = updated_book_loans['date_of_return'].apply(serial_date_to_datetime)

# Calculate total number of loans
total_loans = len(updated_book_loans)

# Calculate total number of days borrowed, ensure no negative days are counted
updated_book_loans['days_borrowed'] = (updated_book_loans['date_of_return'] - updated_book_loans['date_of_loan']).dt.days
updated_book_loans['days_borrowed'] = updated_book_loans['days_borrowed'].apply(lambda x: max(x, 0))
total_days_borrowed = updated_book_loans['days_borrowed'].sum()

# Calculate the average number of days a book was borrowed
average_days_borrowed = total_days_borrowed / total_loans if total_loans > 0 else 0

# Output the results
print("\033[1mAverage days a book was borrowed:\033[0m" ,average_days_borrowed)


Average days a book was borrowed: 10.020371208691715


it shows How many days a each book have been minimum borrowed and it shows the average of duration of all books in library.

## **b) The percentage proportion of books returned late.**

### ***Count of all loan transactions. Computes the days each book was borrowed. Determines if the loan was returned late by checking if days_borrowed is greater than 14. Calculates the proportion of late returns relative to the total number of loans and formats it as a percentage.***

In [ ]:
import pandas as pd
import datetime

# Define your header names as a list
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Read the CSV file with the specified header names
updated_book_loans = pd.read_csv('/content/bookloans (1).csv', header=None, names=header_names)

# Define the function to convert Excel serial date to datetime
def serial_date_to_datetime(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=s)
    return excel_epoch + delta

# Apply conversion to the loan and return dates
updated_book_loans['date_of_loan'] = updated_book_loans['date_of_loan'].apply(serial_date_to_datetime)
updated_book_loans['date_of_return'] = updated_book_loans['date_of_return'].apply(serial_date_to_datetime)

# Calculate total number of loans
total_loans = len(updated_book_loans)

# Calculate total number of days borrowed, ensure no negative days are counted
updated_book_loans['days_borrowed'] = (updated_book_loans['date_of_return'] - updated_book_loans['date_of_loan']).dt.days
updated_book_loans['days_borrowed'] = updated_book_loans['days_borrowed'].apply(lambda x: max(x, 0))

# Calculate number of loans returned late
updated_book_loans['is_late'] = updated_book_loans['days_borrowed'] > 14
total_books_returned_late = updated_book_loans['is_late'].sum()

# Calculate the percentage proportion of books returned late
percentage_late = (total_books_returned_late / total_loans * 100) if total_loans > 0 else 0

# Output the result
print("\033[1mPercentage of books returned late:\033[0m"   f"  {percentage_late:.2f}%")


Percentage of books returned late:  27.75%


The result shows in average percentage that from library how many books are returned late. The 27.75% books from total books are returned late in library.

## **c) The average late period of books returned late.**

### ***calculate the average late period of books returned late. determine how many days late each book was, and then calculate the average of these days.***

In [ ]:
import pandas as pd
import datetime

# Define your header names as a list
header_names = ["book_number", "member_number", "date_of_loan", "date_of_return"]

# Read the CSV file with the specified header names
updated_book_loans = pd.read_csv('/content/bookloans (1).csv', header=None, names=header_names)

# Define the function to convert Excel serial date to datetime
def serial_date_to_datetime(s):
    excel_epoch = datetime.datetime(1899, 12, 30)  # Adjustment for Excel's leap year bug
    delta = datetime.timedelta(days=s)
    return excel_epoch + delta

# Apply conversion to the loan and return dates
updated_book_loans['date_of_loan'] = updated_book_loans['date_of_loan'].apply(serial_date_to_datetime)
updated_book_loans['date_of_return'] = updated_book_loans['date_of_return'].apply(serial_date_to_datetime)

# Calculate total number of loans
total_loans = len(updated_book_loans)

# Calculate total number of days borrowed, ensure no negative days are counted
updated_book_loans['days_borrowed'] = (updated_book_loans['date_of_return'] - updated_book_loans['date_of_loan']).dt.days
updated_book_loans['days_borrowed'] = updated_book_loans['days_borrowed'].apply(lambda x: max(x, 0))

# Calculate number of loans returned late
updated_book_loans['days_late'] = updated_book_loans['days_borrowed'] - 14
updated_book_loans['days_late'] = updated_book_loans['days_late'].apply(lambda x: x if x > 0 else 0)
total_days_late = updated_book_loans['days_late'].sum()

# Calculate total number of late returns
updated_book_loans['is_late'] = updated_book_loans['days_late'] > 0
total_books_returned_late = updated_book_loans['is_late'].sum()

# Calculate the average late period of books returned late
average_late_days = (total_days_late / total_books_returned_late) if total_books_returned_late > 0 else 0

# Output the results
print(f" \033[1mbooks returned late:\033[0m  {total_books_returned_late:.0f}")  # Number of loans returned late
print(f" \033[1mAverage:\033[0m              {average_late_days:.1f}")  # Average late period of books returned late


 books returned late:  613
 Average:              3.5


Out of the books that were brought back late, they were typically delayed by about 3.5 days on average. So, if a book was returned late, it was usually about 3.5 days overdue. And total late days of overall book is 613.  

# **Comment on data**

Markdown is beneficial for organizing and documenting code, providing clarity through formatting. It helps me structure explanations, outline code sections, and highlight important details for better readability. However, the data files provided lack in headers and proper database format, making it challenging to extract information efficiently. This increases the complexity of data processing and querying tasks. Despite these limitations. I implement functions for genre popularity, loan statistics, and late returns. I comment on the code for clarit. Markdown helps in presenting my approach and insights effectively, facilitating understanding for both myself and assessors.